### PR TITLE
Adopt the new aj match condition expression

### LINKED ISSUE #3560: Change aj/raj in Python API to match the recent changes in the Java API.
In Java/Groovy, aj/raj have been improved to support join expressions other than equality. 
```
j2 = myTrades.aj(t, ["Timestamp<Timestamp"], ["PriceBeforeMyTrade=Price"]).view(["Sym", "Price", "PriceBeforeMyTrade", "Timestamp"])
```

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3560

### CODE REVIEW

File: py/server/deephaven/table.py [hunk 1]
```diff
@@ -474,13 +474,13 @@ class SortDirection(Enum):
...
-    LESS_THAN = _JAsOfMatchRule.LESS_THAN
-    GREATER_THAN_EQUAL = _JAsOfMatchRule.GREATER_THAN_EQUAL
-    GREATER_THAN = _JAsOfMatchRule.GREATER_THAN
+# class AsOfMatchRule(Enum):
+#     """An enum defining matching rules on the final column to match by in as-of join and reverse as-of join
+#     operation. """
+#     LESS_THAN_EQUAL = _JAsOfMatchRule.LESS_THAN_EQUAL
+#     LESS_THAN = _JAsOfMatchRule.LESS_THAN
+#     GREATER_THAN_EQUAL = _JAsOfMatchRule.GREATER_THAN_EQUAL
+#     GREATER_THAN = _JAsOfMatchRule.GREATER_THAN
```
Comment (devinrsmith) [inline py/server/deephaven/table.py:483]: Can probably delete this?

File: py/server/deephaven/table.py [hunk 2]
```diff
@@ -1271,46 +1270,38 @@ def aj(self, table: Table, on: Union[str, Sequence[str]], joins: Union[str, Sequ
 
         Args:
             table (Table): the right-table of the join
-            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or an equal expression,
-                i.e. "col_a = col_b" for different column names
+            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
+                two columns, e.g. 'col_a < col_b'. The other valid comparison operator is '<='.
```
Comment (devinrsmith) [inline py/server/deephaven/table.py:1274]: Only the _last_ column can have a non-equals comparison operator:

```python
# these are all less-than-or-equal on baz
["foo", "bar", "baz"]
["foo", "bar", "baz=baz"]
["foo", "bar", "baz==baz"]
["foo", "bar", "baz<=baz"]

# this is the only logically different one
["foo", "bar", "baz<baz"]
```

File: py/server/deephaven/table.py [hunk 3]
```diff
@@ -1271,46 +1270,38 @@ def aj(self, table: Table, on: Union[str, Sequence[str]], joins: Union[str, Sequ
...
         row values equal the row values from the right table where the keys from the left table most closely match
         the keys from the right table without going under. If there is no matching key in the right table, appended row
         values are NULL.
 
         Args:
             table (Table): the right-table of the join
-            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or an equal expression,
-                i.e. "col_a = col_b" for different column names
+            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
+                two columns, e.g. 'col_a > col_b'. The other valid comparison operator is '>='.
```
Comment (devinrsmith) [inline py/server/deephaven/table.py:1302]: Similar comment about the last column.

**Review by rachelmbrubaker — COMMENTED** [review]
LGTM thanks 😄

File: py/server/deephaven/table.py (hunk 1 above)
Comment (chipkent) [inline py/server/deephaven/table.py:483]: remove

File: py/server/deephaven/table.py (hunk 2 above)
Comment (chipkent) [reply py/server/deephaven/table.py:1274]: I agree that what is here is not clear.

File: py/server/deephaven/table.py (hunk 3 above)
Comment (chipkent) [reply py/server/deephaven/table.py:1302]: yep

File: py/server/deephaven/table.py [hunk 4]
```diff
@@ -2901,44 +2886,29 @@ def aj(self, table: Union[Table, PartitionedTableProxy], on: Union[str, Sequence
 
         Args:
             table (Union[Table, PartitionedTableProxy]): the right table or PartitionedTableProxy of the join
-            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or an equal expression,
-                i.e. "col_a = col_b" for different column names
+            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
+                two columns, e.g. 'col_a < col_b'. The other valid comparison operator is '<='.
```
Comment (chipkent) [inline py/server/deephaven/table.py:2890]: Here too

File: py/server/deephaven/table.py [hunk 5]
```diff
@@ -2948,38 +2918,24 @@ def raj(self, table: Union[Table, PartitionedTableProxy], on: Union[str, Sequenc
 
         Args:
             table (Union[Table, PartitionedTableProxy]): the right table or PartitionedTableProxy of the join
-            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or an equal expression,
-                i.e. "col_a = col_b" for different column names
+            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
+                two columns, e.g. 'col_a > col_b'. The other valid comparison operator is '>='.
```
Comment (chipkent) [inline py/server/deephaven/table.py:2922]: here too

File: py/server/tests/test_pt_proxy.py [hunk 6]
```diff
@@ -209,14 +209,14 @@ def test_as_of_join(self):
                 joined_pt_proxy = pt_proxy.aj(right_table, on=["a"])
                 self.assertTrue([ct for ct in joined_pt_proxy.target.constituent_tables if ct.size > 0])
 
-                joined_pt_proxy = pt_proxy.aj(right_table, on=["a"], joins="e", match_rule=AsOfMatchRule.LESS_THAN)
+                joined_pt_proxy = pt_proxy.aj(right_table, on=["a < a"], joins="e")
```
Comment (chipkent) [inline py/server/tests/test_pt_proxy.py:212]: @jjbrosnan @margaretkennedy This PR is a breaking change.  If the last column is named the same, you get expressions like these, which users are going to find weird.  We need to carefully address this in our docs.

File: py/server/deephaven/table.py (hunk 2 above)
Comment (chipkent) [reply py/server/deephaven/table.py:1274]:
```suggestion
            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
                two columns, e.g. 'col_a = col_b'. The first 'N-1' matches are exact matches.  The final match is an inexact match.  The inexact match can use either '<' or '<='.  If a common name is used for the inexact match, '<=' is used for the comparison.
```

File: py/server/deephaven/table.py (hunk 3 above)
Comment (chipkent) [reply py/server/deephaven/table.py:1302]:
```suggestion
            on (Union[str, Sequence[str]]): the column(s) to match, can be a common name or a match condition of
                two columns, e.g. 'col_a = col_b'. The first 'N-1' matches are exact matches.  The final match is an inexact match.  The inexact match can use either '>' or '>='.  If a common name is used for the inexact match, '>=' is used for the comparison.
```

**Review by chipkent — CHANGES_REQUESTED** [review]

**Review by chipkent — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/2418
Conceptual: https://github.com/deephaven/deephaven.io/issues/2417
Reference: https://github.com/deephaven/deephaven.io/issues/2416
Blog: https://github.com/deephaven/deephaven.io/issues/2415

**Pull request merged**